### PR TITLE
[device] return deviceType as a constant, new implementation on android

### PR DESCRIPTION
**Pull request by maxencehenneron**
# Why

getDeviceTypeAsync doesn't need to be async and can be a constant, iOS, Android and Web all use synchronous code. I kept the method for backward compatibility

On Android, the function was not returning the correct value on many devices. ( from our customer support: OnePlus Nord N10, Motorola G7 and Nokia x100 all return DeviceType.TABLET when they clearly are phone) 
The official [android documentation](https://developer.android.com/guide/topics/large-screens/support-different-screen-sizes) suggests that 99.96% of phones have a width of less than 600dp so I suggest using that to determine the device type on Android. This works with the devices mentionned above. (Implemented in this PR).

I added all the tests

# How

I exported the deviceType as a constant.

# Test Plan

Run the test suite, run the app on tablet and navigate to the "Device screen". A new deviceType (Constant) section has been added.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.kt [hunk 1]
```diff
@@ -151,23 +153,32 @@ class DeviceModule : Module() {
...
-        DeviceType.UNKNOWN
+        // Devices larger than MIN_TABLET_SIZE_DP are considered tablets
+        DeviceType.TABLET
       }
     }
+
+    private fun dpiFromPx(size: Int, densityDpi: Int): Float {
+      val densityRatio = densityDpi.toFloat() / DisplayMetrics.DENSITY_DEFAULT
+      return size / densityRatio
+    }
```
Comment (ide) [inline packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.kt:182]: Android docs suggest never to implement this equation and to use https://developer.android.com/reference/android/util/TypedValue#applyDimension(int,%20float,%20android.util.DisplayMetrics) instead (PX to DIP in this case).

**Review by ide — CHANGES_REQUESTED** [review]
Could you please make this a sync function instead of a property since computing the value performs "work"?

**Comment by maxencehenneron** [comment]
> Could you please make this a sync function instead of a property since computing the value performs "work"?

I thought about this but making this a sync function would do "work" everytime we call the function, while making it a constant would only do the work when initially loading the module. I believe this work is also negligible as we're only reading properties from windowManager on Android and on iOS we're just mapping UIDevice.current.userInterfaceIdiom.

This constant may be used a lot. Here's the use case:

I currently use react-native-device-info which I believe was initially used as an inspiration to this module due to the many lines of code they have in common. 
react-native-device-info now has a "isTablet" and "deviceType" constant that can be used to render the view differently on tablets, phones and desktop without being dependant on the component state so the styles can be extracted from the component.

https://github.com/react-native-device-info/react-native-device-info/blob/f0bf711ead00e50fe8b576ffa44c542012e06358/android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java#L202

For example, in this file in my source code, I'm accessing this constant quite a lot. Making this a function would do the same work multiple times in all my stylesheets. I could create a context to store the value but that would make all my stylesheets dependant on this context.

Screen Shot 2023-01-05 at 9 50 03 AM

What are your thoughts on this? 

I agree with your second comment and will update my code to use `applyDimension`.

**Comment by ide** [comment]
I talked to @tsapeta and he says he has plans to support lazily computed properties, which will help with the performance issue. Generally for a JS API I would like to avoid putting behavior behind properties and prefer functions over getters for idiomatic reasons, unlike a getter-heavy language like Kotlin or C#. But you have a clear use case that would help you build more libraries on top of expo-device so let's go with the property here.

**Comment by tsapeta** [comment]
I'm closing the PR then, but feel free to let us know here or open a new PR whenever you want to continue working on that 😉

**Pull request closed without merging**